tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/TeamMailboxMemberSetMethod.scala

```diff
@@ -0,0 +1,121 @@
+        .filter(userExist => !userExist)
+        .map(_ => username))
+
+  private def updateTeamMailboxMembers(username: Username,
+                                       teamMailbox: TeamMailbox,
+                                       membersUpdateToAdd: List[TeamMailboxMember],
+                                       membersUpdateToRemove: Set[Username]): SMono[TeamMailboxMemberSetResult] =
+    SFlux(teamMailboxRepository.listMembers(teamMailbox))
+      .collectMap(member => member.username, member => member)
+      .flatMap(presentMembers => presentMembers.get(username) match {
+        case Some(member) =>
+          member.role.value match {
+            case ManagerRole => updateTeamMailboxMembers(teamMailbox, membersUpdateToAdd, membersUpdateToRemove, presentMembers)
+            case _ => SMono.just(TeamMailboxMemberSetResult.notUpdated(teamMailbox, SetErrorDescription(s"Not manager of teamMailbox ${teamMailbox.asString()}")))
+          }
+        case None => SMono.just(TeamMailboxMemberSetResult.notUpdated(teamMailbox, SetErrorDescription("Team mailbox is not found or not a member of the mailbox")))
+      })
+
+  private def updateTeamMailboxMembers(teamMailbox: TeamMailbox,
+                                       membersUpdateToAdd: List[TeamMailboxMember],
+                                       membersUpdateToRemove: Set[Username],
+                                       presentMembers: Map[Username, TeamMailboxMember]): SMono[TeamMailboxMemberSetResult] =
+    SFlux.fromIterable(membersUpdateToRemove.concat(membersUpdateToAdd.map(teamMailboxMember => teamMailboxMember.username)))
+      .filter(username => presentMembers.contains(username) && ManagerRole.equals(presentMembers(username).role.value))
+      .next()
+      .map(username => TeamMailboxMemberSetResult.notUpdated(teamMailbox, SetErrorDescription(s"Could not update or remove manager ${username.asString()}")))
+      .switchIfEmpty(updateMembers(teamMailbox, membersUpdateToAdd, membersUpdateToRemove))
+
+  private def updateMembers(teamMailbox: TeamMailbox,
+                            membersUpdateToAdd: List[TeamMailboxMember],
+                            membersUpdateToRemove: Set[Username]): SMono[TeamMailboxMemberSetResult] =
+    removeMember(teamMailbox, membersUpdateToRemove)
+      .`then`(addMember(teamMailbox, membersUpdateToAdd))
+      .`then`(SMono.just(TeamMailboxMemberSetResult(Option[TeamMailboxNameDTO](TeamMailboxNameDTO(teamMailbox.asString())), Option.empty)))
+
+  private def addMember(teamMailbox: TeamMailbox, members: List[TeamMailboxMember]): SMono[Unit] =
+    SFlux.fromIterable(members)
+      .flatMap(member => teamMailboxRepository.addMember(teamMailbox, member))
+      .`then`()
+
+  private def removeMember(teamMailbox: TeamMailbox, usernames: Set[Username]): SMono[Unit] =
+    SFlux.fromIterable(usernames)
+      .flatMap(username => teamMailboxRepository.removeMember(teamMailbox, username))
```

Comment:
Feeling sad that we actually end up splitting our updates into individual inserts / remove.
Appart from generating more requests we loose atomicity of the requests.
That's sad because the underlying layers (mailbox ACL) do support full/partial updates in an atomic fashion.
A well designed reposiotry layer would have had exposed this inner structure and allowed to leverage atomicity onto the presentation layer.

Reply:
Move this issue to another ticket https://github.com/linagora/james-project/issues/5221

Reply:
From what I could read you are about to attempt a massive reset of all the rights, which is not idempotent and even a regression.

Maybe we cannot achieve full atomicity.

I thought the ACL command was multi-valued (one command -> several keys)

It is actually not multi-valued so existing mechanisms do not add benefits.

Sorry for bothering.